### PR TITLE
fixed date time zone

### PR DESCRIPTION
**Pull request by OlyaAn**
[AB#59428](https://dev.azure.com/orckestra001/da965986-c85b-43be-9915-219568b000e5/_workitems/edit/59428)

### CODE REVIEW

**Review by TONYFECTEAU — APPROVED** [review]

File: src/components/MaterialUI/Inputs/DatePicker.js [hunk 1]
```diff
@@ -133,10 +137,25 @@ const WrappedDatePicker = ({
...
+	const localizedTimeZoneName = useSelector(namedLookupLocalizedSelector("customer", "TimeZone", timePickerTimeZone));
+	const setLocalZone = (date, timezone) => {
+		const dateWithoutZone = moment.tz(date, timezone).format("YYYY-MM-DDTHH:mm:ss.SSS");
+		const localZone = moment(dateWithoutZone).format("Z");
+		const dateWithLocalZone = [dateWithoutZone, localZone].join("");
+		return new Date(dateWithLocalZone);
+	}
+	  
+	const setOtherZone = (date, timezone) => {
+		console.log(timezone);
```
Comment (InnaBoitsun) [inline src/components/MaterialUI/Inputs/DatePicker.js:149]: remove console.log
and move this methods to timezoneHelper and cover with tests

File: src/components/MaterialUI/Inputs/DatePicker.js [hunk 2]
```diff
@@ -147,12 +166,12 @@ const WrappedDatePicker = ({
 					<DatePicker
 						{...props}
 						dateFormat={dateFormat || createFormat(useDate, useTime)}
-						selected={startDate}
-						onChange={date => updateDate(date, metadata)}
+						selected={startDate ? setLocalZone(startDate, getTimeZoneByName(timePickerTimeZone)) : null}
```
Comment (InnaBoitsun) [inline src/components/MaterialUI/Inputs/DatePicker.js:169]: set correct value where the startDate calculated, on line #138

File: src/components/MaterialUI/Inputs/DatePicker.js [hunk 3]
```diff
@@ -133,10 +137,25 @@ const WrappedDatePicker = ({
...
+		const otherZone = moment.tz(date, timezone).format("Z");
+		const dateWithOtherZone = [dateWithoutZone, otherZone].join("");
+		return new Date(dateWithOtherZone);
+	}
+	
 	const updateDate = (date, metadata) => {
 		if (onChange) {
-			onChange(date, metadata);
+			onChange(date ? setOtherZone(date, getTimeZoneByName(timePickerTimeZone)) : null, metadata)
 		}
```
Comment (InnaBoitsun) [inline src/components/MaterialUI/Inputs/DatePicker.js:159]: overwrite local timezone just when custom timezone is specified and time is used for picker, not always

**Review by InnaBoitsun — CHANGES_REQUESTED** [review]

File: src/components/MaterialUI/Inputs/DatePicker.js [hunk 4]
```diff
@@ -131,12 +134,18 @@ const WrappedDatePicker = ({
 	...props
 }) => {
 	const classes = useStyles({ readOnly });
-	const startDate = value ? new Date(value) : null;
+	const startDate = value ? setLocalZoneDate(new Date(value), getTimeZoneByName(timePickerTimeZone)) : null;
```
Comment (InnaBoitsun) [inline src/components/MaterialUI/Inputs/DatePicker.js:137]: define and calculate timeZoneName once and then use
const timeZoneName = getTimeZoneByName(timePickerTimeZone);

File: src/utils/timezoneHelper.js [hunk 5]
```diff
@@ -14,6 +16,20 @@ export const getTimeZoneName = () => {
...
 };
 
+export const setLocalZoneDate = (date, timezone) => {
+	const dateWithoutZone = moment.tz(date, timezone).format("YYYY-MM-DDTHH:mm:ss.SSS");
+    const localZone = moment(dateWithoutZone).format("Z");
+    const dateWithLocalZone = [dateWithoutZone, localZone].join("");
+	return new Date(dateWithLocalZone);
+}
+
+export const setOtherZoneDate = (date, timezone) => {
```
Comment (InnaBoitsun) [inline src/utils/timezoneHelper.js:26]: let rename to convertTimeToOtherTimeZone    and add some sammary

File: src/utils/timezoneHelper.js [hunk 6]
```diff
@@ -14,6 +16,20 @@ export const getTimeZoneName = () => {
 	return timezoneName;
 };
 
+export const setLocalZoneDate = (date, timezone) => {
```
Comment (InnaBoitsun) [inline src/utils/timezoneHelper.js:19]: et rename to convertTimeToLocalTimeZone and add some summary

File: src/components/MaterialUI/Inputs/DatePicker.js [hunk 7]
```diff
@@ -131,12 +138,18 @@ const WrappedDatePicker = ({
...
+	const localizedTimeZoneName = useSelector(namedLookupLocalizedSelector("customer", "TimeZone", timePickerTimeZone));
 
 	const updateDate = (date, metadata) => {
 		if (onChange) {
-			onChange(date, metadata);
+			if (useTime && timePickerTimeZone) {
+				onChange(date ? convertTimeToOtherTimeZone(date, getTimeZoneByName(timePickerTimeZone)) : null, metadata);
+			} else {
+				onChange(date, metadata);
+			}
```
Comment (InnaBoitsun) [inline src/components/MaterialUI/Inputs/DatePicker.js:152]: can be simplified like this?
```
if (onChange) {
   onChange((useTime && timePickerTimeZone) ? convertTimeToOtherTimeZone(date, timeZoneName) : date, metadata);
}
```

File: src/components/MaterialUI/Inputs/DatePicker.js [hunk 8]
```diff
@@ -131,12 +138,18 @@ const WrappedDatePicker = ({
 	...props
 }) => {
 	const classes = useStyles({ readOnly });
-	const startDate = value ? new Date(value) : null;
+	const timeZoneName = getTimeZoneByName(timePickerTimeZone);
+	const startDate = value ? convertTimeToLocalTimeZone(new Date(value), timeZoneName) : null;
```
Comment (InnaBoitsun) [inline src/components/MaterialUI/Inputs/DatePicker.js:142]: do we need to convert when timePickerTimeZone is undefined?

File: src/components/MaterialUI/Inputs/DatePicker.test.js [hunk 9]
```diff
@@ -311,6 +328,130 @@ describe("DatePicker", () => {
...
+			</TestWrapper>
+		);
+		const mountedComponent = mount(component);
+
+		const input = mountedComponent.find("input");
+		input.at(0).simulate("change", event);
+	});
+
+	it("should call onChange prop with timezone without value", () => {
+		const onChangeMock = jest.fn();
```
Comment (InnaBoitsun) [inline src/components/MaterialUI/Inputs/DatePicker.test.js:430]: is the test name corresponding what it is doing?

File: src/components/MaterialUI/Inputs/DatePicker.test.js [hunk 10]
```diff
@@ -311,6 +328,130 @@ describe("DatePicker", () => {
...
+			</TestWrapper>
+		);
+		const mountedComponent = mount(component);
+
+		const input = mountedComponent.find("input");
+		input.at(0).simulate("change", event);
+		expect(onChangeMock.mock.calls.length, "to equal", 1);
+	});
+
+	it("should call onChange prop with timezone without value", () => {
```
Comment (InnaBoitsun) [inline src/components/MaterialUI/Inputs/DatePicker.test.js:387]: is the test name corresponding what it is doing?

**Review by InnaBoitsun — APPROVED** [review]

**Review by InnaBoitsun — APPROVED** [review]

**Pull request merged**